Update to Dogma 0.2.0

jmalloc opened the pull request:
This change updates Dogma version to 0.2.0, which means the memory engine needs to  implement the `Key()` and `Time()` methods to `ProjectionEventScope`.

In order to implement `Time()`, I had to add a `Time` field to `envelope.Envelope`, which necessitated some fairly wide-reaching changes through the controller implementations.

koden-km commented on engine/envelope/envelope.go:
```diff
@@ -78,17 +85,19 @@ func (e *Envelope) NewTimeout(
 	id string,
 	m dogma.Message,
 	t time.Time,
+	s time.Time,
```
What is `t` and `s`? Due to there being two time values, should the docs describe these a bit?

jmalloc replied:
Yeah, you're right.

I think I'll rename the struct fields too:

- `Time` -> `CreatedAt`
- `TimeoutTime` -> `ScheduledFor`

Or something like that.

koden-km replied:
That sounds much clearer.

jmalloc commented:
Ok, this PR crept in scope a bit - I took it as an opportunity to improve the interface of `Envelope` as it's grown a bit.

- Added new `CreatedAt` time
- Renamed `TimeoutTime` to `ScheduledFor` and also change it from a `*time.Time` to just `time.Time`
- Ditched the `envelope.New()` function in favour of explicit `NewCommand()` and `NewEvent()` functions (to avoid passing roles around in the API).
